### PR TITLE
Starting the Shark Contract

### PR DESCRIPTION
**Pull request by FliqHydro**

### CODE REVIEW

**Comment by raysapida** [comment]
@FliqHydro Testing notification

File: contracts/Shark.sol [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    string[10] public species;
+    address[] public owners;
+    struct SharkObject {
+        uint speed;
+        uint strength;
+        string type;
+    }
+
+    function createBasicShark() public {
+        owners[SharkObject(10, 10, "black tip reef shark")] = msg.sender;
```
Comment (raysapida) [inline contracts/Shark.sol]: @FliqHydro check syntax

**Comment by FliqHydro** [comment]
@raysapida Hello

**Comment by FliqHydro** [comment]
@raysapida I'm having an issue, and can't even migrate to start the Game contract. Sorry about this being so late. Here's the error. ianwh@DESKTOP-USVPDCN MINGW64 /C/Users/ianwh/Desktop/Blockchain (contracts)
$ truffle migrate
Error parsing C:\Users\ianwh\Desktop\Blockchain\contracts\Shark.sol: ParsedContract.sol:7:5: ParserError: Expected token Semicolon got 'Struct'
    struct SharkObject; {
    ^
Compilation failed. See above.
I can't figure it out, but its probably something stupid. IF you get back to me tonight then I will try to Get started on the game contract.

**Comment by raysapida** [comment]
@FliqHydro It's alright, I think it's missing a semicolon at the end of line 6. Also, you might need to rename the array on line 5 to `sharks` instead of `owners` so that it won't interfere with mapping.

**Comment by FliqHydro** [comment]
@raysapida I have come across another error, I've tried some things, but this is where you said you weren't sure about the syntax. //TODO: Double check struct syntax
    function getShark() public view returns (struct) {

Error: ianwh@DESKTOP-USVPDCN MINGW64 /C/Users/ianwh/Desktop/Blockchain (contracts)
$ truffle migrate
Error parsing C:\Users\ianwh\Desktop\Blockchain\contracts\Shark.sol: ParsedContract.sol:22:46: ParserError: Expected type name
    function getShark() public view returns (struct) {
                                             ^
Compilation failed. See above.

**Comment by raysapida** [comment]
@FliqHydro Ahh, I see. I refactored the Shark contract below based on the CryptoZombies website; we can go over the `getSharks` method next time we pair program though. Have you seen for loops before? 

```
pragma solidity ^0.4.18;

contract Shark {
  string[10] public species;
  SharkObject[] public sharks;
  mapping (uint => address) owners;
  mapping (address => uint) ownerSharkCount;

  struct SharkObject {
    uint speed;
    uint strength;
    string typeOfShark;
  }

  function createBasicShark() public {
    uint id = sharks.push(SharkObject(10, 10, species[0])) - 1;
    owners[id] = msg.sender;
    ownerSharkCount[msg.sender]++;
  }

  // TODO: Separate speed and strength
  function trainShark(uint _sharkId) public {
    require(msg.sender == owners[_sharkId]);
    sharks[_sharkId].speed ++;
    sharks[_sharkId].strength ++;
  }

  function getSharks() public view returns(uint[]) {
    uint[] memory result = new uint[](ownerSharkCount[msg.sender]);
    uint counter = 0;
    for (uint i = 0; i < sharks.length; i++) {
      if (owners[i] == msg.sender) {
        result[counter] = i;
        counter++;
      }
    }
    return result;
  }

}
```

**Comment by FliqHydro** [comment]
No I have not seen loops before

File: contracts/SharkRace.sol [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    return uint(keccak256(now, msg.sender, randNonce)) % _modulus;
+  }
+
+  function sharkRace(uint _sharkId, uint _targetId) external  {
+    require (msg.sender == owners[_sharkId]);
+    SharkObject storage myShark = sharks[_sharkId];
+    SharkObject storage enemyShark = sharks[_targetId];
+    uint rand = randMod(100);
+    if (rand <= attackVictoryProbability) {
+//myShark.winCount++;
```
Comment (raysapida) [inline contracts/SharkRace.sol]: To use `winCount`, `level`, and `lossCount`, you'll have to add those fields in the `SharkObject` in the `SharkContract`

File: contracts/Shark.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+contract Shark {
+    string[10] public species;
+    SharkObject[] public sharks;
+    mapping (uint => address) owners;
+    mapping (address=> uint) ownerSharkCount;
+    struct SharkObject {
+        uint speed;
+        uint strength;
+        string typeOfShark;
```
Comment (raysapida) [inline contracts/Shark.sol:11]: You can add `winCount`, `level`, and `lossCount` after this line. All three should be numbers so using `uint` would still work.

**Comment by raysapida** [comment]
If you wanted to get started with adding how the Sharks would transfer ownership or to auction them, this contract https://github.com/raysapida/zombie-game/blob/master/contracts/ZombieOwnership.sol
is a good starting point. It might be too tough to tackle on your own though so you also could add this contract https://github.com/raysapida/zombie-game/blob/master/contracts/erc721.sol to your project and then we could talk about it next time we pair program

**Pull request merged**